### ISSUE TITLE
Unipept Shotgun Metagenomics Pipeline

### ISSUE THREAD

**Issue by pdawyndt** [unipept/unipept#316]
**planning**

before new year
- in memory index to speed up Unipept database construction from Uniprot + NCBI Taxonomy
- use new indexing in Unipept database construction scripts
- ship rails platform in docker
- think about different ways of shipping the database for runnig multiple instances
- benchmarks
  - querying: mySQL index vs berkleyDB index
  - querying: 1 or multiple instances of index
  - multiple instances: variation in number of instances vs available RAM
- sharding (slicing the database content over multiple machines)
- k-mer indexing (see Kraken)

**input**
- list of protein reads (partial proteins) identified from an environmental sample
- list of DNA reads from a shotgun metagenomics pipeline; gene calling software (eg. FragGeneScan) is used to convert this input to the above input

**output**
- [ ] taxonomic assignments for each of the protein/DNA reads in the data set
- [ ] graphical representation of the distribution of taxonomic assignments across the data set

**algorithmic outline**
- build index structure for the NCBI/Unipept Taxonomy that allows
  - fast lookup of the lineage for a given taxon ID
  - fast computation of the lowest common ancestor (LCA) for a given list of taxa, or any other taxonomic aggregation (#266) for a given list of (weighted) taxa
  - composition of multiple taxonomies (#266)
  - data structure should be read in memory for increased performance, and should be usefull both as a component of the metagenomics pipeline, as a component in the Unipept middleware that underpins the Unipept website and the Unipept API, and to run client-side in combination with the Unipept API (web services and command line interface; could be an option for the command line tools to run taxonomic queries either against the Unipept API or against the local data structure)
  - if data structure will only be queried with valid taxon IDs (from the Unipept Taxonomy), invalid leaf nodes from the NCBI Taxonomy Database should not be included (saves time and memory); this should be an option when the datastructure is constructed
- build index structure that can do fast mapping from substrings to their taxonomic assignment
  - use exact full-length matching to match the substrings
  - estimated size for tryptic peptides extracted from Uniprot: mapping keys are about 1 billion tryptic peptides, 5-50 characters in length, 20 character alphabet
  - note: EBI has some good experience in using CouchBase as a NoSQL solution for indexing
- combine above two datastructures into Unipept Shotgun Metagenomics Pipeline
  - precompute taxonomic assignments for each of the possible substring found in a reference dataset
    - LCA (or other taxonomic assignment) for each tryptic peptide in Uniprot
    - LCA (or other taxonomic assignment) for each k-mer in the RefSeq genomes (following Wood & Salzberg (2014))
    - LCA (or other taxonomic assignment) for each k-mer in Uniprot (this would give Unipept an index that can do general substring matching, not just full-length tryptic peptide matching)
  - split input protein/DNA read into several substrings according to a fixed scheme
    - tryptic peptides in the Unipept approach (note that peptides at the start and end of the protein fragment may be discarded, as the likelihood that they will be tryptic peptides is quite low)
    - overlapping k-mers in the [Kraken](http://genomebiology.com/2014/15/3/R46) approach
    - deduplicate list of substrings into set of substrings before mapping them to their taxonomic assignments, in order to avoid duplicate work (build frequency table so that taxonomic assignments can be weighted, or lookup table so that substrings can be aggregated at the protein/DNA read level); deduplication may also happen across all samples for a given project
  - map each substring onto its taxonomic assignment
  - aggregate taxonomic assignments for all substrings into a taxonomic assignment on the protein/DNA read level (optional); for paired-end reads, taxonomic assignments for a pair of DNA reads can be aggregated at the level of the pair, since all assignments in either end of the pair come from the same organism
  - aggregate taxonomic assignments for all substrings, protein reads/DNA reads on the sample level
  - visualize aggregated taxonomic assignments at the sample level
  - compare aggregated taxonomic assignments from multiple samples
- index construction
  - possibility to quickly store and retrieve the index structure to disk
  - `unipept` command should be able to call the Unipept API or some local instances of the resources (this would allow power-user to run one or more local instances of the index and also given them the benefit that they don't have to send their data out)
  - try in-memory construction and parallelization of index structure construction and LCA computations to speed up the time needed to build the index
  - make index persistent in database
  - apart from indexing only the tryptic peptides, we could also index all k-mers for a given k ([Wood & Salzberg 2014](http://genomebiology.com/2014/15/3/R46)); this would give Unipept a more flexible index; protein reads can then be matched using a seed-and-aggregate approach, with aggregation as discussed in #266; note that this seed-and-aggregate approach follows the MapReduce programming model for processing large data sets in parallel; in the mapping phase, the lowest common ancestor (or any other aggregation) is computed/looked up for a series of matching seeds (tryptic peptides, overlapping k-mers, ...) and in the reduce phase, the lowest common ancestors are aggregated into a consensus biodiversity; here's the general mapreduce model

![usmp_mapreduce](https://github.ugent.be/github-enterprise-assets/0000/0006/0000/0172/0f755a26-de87-11e3-9f70-bb4d265a0fea.png)

**targetted assets**
- should be as fast a possible, and possibly also memory efficient so that it can be run on a modern laptop computer (let's say 4Gb RAM)
- compare against largest possible reference database (e.g. Uniprot or M5nr)
- identify all kingdoms of life simultaneously: _Archaea_, _Bacteria_, _Eukaryotes_ and _Viruses_
- supplied as a virtual machine that can run in the cloud (#310)

**processed metagenomics sample: interesting statistics**
- number of mapped substrings
- number of strings mapped to root LCA
- number of unmapped substrings
- number of mapped protein/DNA reads
- number of unmapped protein/DNA reads
- average number of substrings per protein/DNA read
- average number of mapped substrings per protein/DNA read
- number of protein/DNA read level aggregated taxonomic assignments that result in a single lineage (no forks) if hierarchical aggregation is used; also show these results per number of mapped substrings per protein/DNA read, as it is logical that all proteins or DNA reads for which there's only a single mapped substring, will automatically result in a single lineage
- distribution of taxonomic ranks for all substrings in the sample
- distribution of taxonomic ranks for all protein/DNA reads in the sample

As processing progresses, these statistics might be appended to an HTML document that user can inspect to follow up processing and to analyse the basic results afterwards.

**literature**
- [Wooley JC, Godzik A, Friedberg I (2010) A Primer on Metagenomics. PLOS Computational Biology](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2829047/)
- [Thomas T, Gilbert J, Meyer F (2012). Metagenomics - a guide from sampling to data analysis. Microbial Informatics and Experimentation 2(1), 3](http://www.ncbi.nlm.nih.gov/pubmed/22587947)
- [MG-RAST user manual](ftp://ftp.metagenomics.anl.gov/data/manual/mg-rast-manual.pdf)
- [Meyer F, Paarmann D, D'Souza M, Olson R, Glass EM, Kubal M, Paczian T, Rodriguez A, Stevens R, Wilke A, Wilkening J, Edwards RA (2008). The Metagenomics RAST server — A public resource for the automatic phylogenetic and functional analysis of metagenomes. BMC Bioinformatics 9, 386.](http://www.biomedcentral.com/1471-2105/9/386)
- [Unipept: Tryptic Peptide-Based Biodiversity Analysis of Metaproteome Samples](http://pubs.acs.org/doi/abs/10.1021/pr300576s)
- [FragGeneScan: predicting genes in short and error-prone reads](http://omics.informatics.indiana.edu/FragGeneScan/)
- [EBI metagenomics – a new resource for the analysis and archiving of metagenomic data](http://nar.oxfordjournals.org/content/42/D1/D600)
- [Wood DE, Salzberg SL (2014). Kraken: ultrafast metagenomic sequence classification using exact alignments. Genome Biology 15, R46](http://genomebiology.com/2014/15/3/R46)
- [Marçais G, Kingsford C: A fast, lock-free approach for efficient parallel counting of occurrences of k-mers. Bioinformatics 2011, 27:764-770.](http://bioinformatics.oxfordjournals.org/content/27/6/764.long)
- protein sequencing ([wikipedia](http://en.wikipedia.org/wiki/Protein_sequencing))

**remarks**
- lots of opportunities for parallelization; draft initial sketch of the metagenomics pipeline and implement prototype to have a good overall idea on what the different components of the pipeline will do and how they will interact
- instead of building a fast map from peptides to LCAs (or other taxonomic aggregation), we could start building two maps: one that maps peptides to proteins and one that maps proteins to taxon IDs (this is what the database is currently used for); the mapping of peptides to LCAs could then be derived from these two mappings; could help in constructing the database (LCA computation)
- since the above is essentially key-value mapping, technologies like in-memory databases and Hadoop could be explored
- test alternatives for LCA computation for individual tryptic peptides (e.g. build probabilistic lineage tree based on counts or occurrences of taxa); can be tested first on reduced datasets since storage of probabilistic lineage trees will be more time-consuming than storage of LCAs; computing both tree types will be quite fast (most time goes to looking up the proteins that match the tryptic peptides), and actually the LCA is just a reduced version of the probabilistic lineage tree
- if LCA is computed for each substring and substrings mapped to the root level have no impact on follow-up analysis, then the final index should not include the substrings mapped to root; since the number of root-mapped-substrings is quite high, this would be a serious reduction in space and time needed to use the index; however, one condition is that we do not want to discriminate between substrings that were not found in the original reference database and substrings that they are so omnipresent in different life forms that we can't learn anything from their source since they are not discriminative enough; another condition is that we operate on LCAs and not a more detailed taxonomic aggregation for these substrings
- to compute the LCAs, we need a mapping from tryptic peptides to a list of taxid (with counts for more complex taxonomic aggregation); for now, this mapping requires a join with the protein table that is the link between peptide and taxon; precomputing this join might speed up retrieving all taxa in which the peptide was found

_[Original issue](https://github.ugent.be/unipept/unipept/issues/316) by @pdawyndt on Tue May 13 2014 at 08:16._

**Comment by pdawyndt** [comment]
A student tried the following approach to compare several metagenomics samples:
- determine metapeptidome of each sample
  - gene calling on DNA reads, resulting in a set of protein sequences for each sample
  - trypsine digest on set of proteins, resulting in a set of tryptic peptides for each sample (sample peptidome or metapeptidome)
- compare metapeptidomes within a population and between populations
  - compute core metapeptidome (intersection of metapeptidomes within a population) 
  - compute unique metapeptidome (intersection of metapeptidomes within a population minus peptidomes of other populations; i.e. what gives a specific profile for lean vs. obese populations)

Although technological issue for metagenomic sequencing are a limiting factor for this approach, this needs further thinking. Here are the initial comments I made on this idea:
- computing the intersection at the level of the tryptic peptides will not work, since we're working with shotgun sequencing data; this means that we just extract a random sample from the whole metagenome (all gut organisms); actually its a random sample of the organisms and on top of that a random sample of short DNA reads of the genomes of these organisms; coverage of metagenomics data sets is too low, so that not the entire genome of all organisms in the sample will be included in the data set; this is the reason why we first map the peptides to the organisms from which they were derived, and then compare the occurrences of the organisms (e.g. by taking the intersection).
- for high-coverage shotgun metagenomics (e.g. if the environment from which the sample was taken is not very rich in biodiversity: only a few species occur) you're approach could definitely work (I hadn't even tought about that), but human gut samples are very rich and current technology does not allow yet the deep sequencing that would be needed for your approach; actually, in your approach you could postpone computing the LCAs until you have determined the intersection of tryptic peptides; this would make this approach extremely fast


_[Original comment](https://github.ugent.be/unipept/unipept/issues/316#issuecomment-9640) by @pdawyndt on Wed May 14 2014 at 08:53._
